### PR TITLE
Fix #6102

### LINKED ISSUE #6102: Relay chain coretime assigner does not support more assignments than fit in a single XCM message (currently 28)
The system allows interlacing right down to the single block level (80 assignments per timeslice, each with a CoreMask with one bit set)
The problem is that it creates a call that doesn't actually fit in an XCM message (we can fit max 28 assignments in a single XCM)

We can easily chunk that on the Coretime Chain side and send it over as four messages, however with the current design that means we need to call `assign_core` multiple times on the relay for a given timeslice [which is disallowed by design](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/runtime/parachains/src/assigner_coretime/mod.rs#L387) due to some assumptions made by the scheduler.

Mitigation in the mean time:
28 assignments is the limit, but 27 assignments that don't add up to a complete mask will be rejected due to the requirement for a full mask on the relay. Therefore we take the first 27 and append an `Idle` assignment, taking it to 28.
This will make anybody who interlaces more than 27 times lose some assignments, but it's better than the current system, which just drops the entire core's assignments because the message is too big. Once this is missed, it's gone from the workplan and is a total mess. Far preferable to truncate and assign everything we can until we can drop some assumptions in the scheduler on the relay.

Mitigation for the Polkadot launch: https://github.com/polkadot-fellows/runtimes/pull/434
Testnets mitigation: https://github.com/paritytech/polkadot-sdk/pull/6022

### PR DESCRIPTION
**Pull request by eskimor**
Relax requirements for `assign_core` so that it accepts updates for the last scheduled entry.

Fixes #6102

### CODE REVIEW

File: polkadot/runtime/parachains/src/assigner_coretime/mod.rs [hunk 1]
```diff
@@ -387,67 +380,56 @@ impl<T: Config> Pallet<T> {
...
-	/// available
+	/// Important only appending is allowed or insertion into the last item is possible. Meaning,
+	/// all already existing assignments must have a begin smaller or equal than the one passed
+	/// here. This restriction exists, because it makes the insertion O(1) and the author could not
+	/// think of a reason, why this restriction should be causing any problems. Inserting
+	/// arbitrarily causes a `DispatchError::DisallowedInsert` error. This restriction could easily
+	/// be lifted if need be and in fact an implementation is available
 	/// [here](https://github.com/paritytech/polkadot-sdk/pull/1694/commits/c0c23b01fd2830910cde92c11960dad12cdff398#diff-0c85a46e448de79a5452395829986ee8747e17a857c27ab624304987d2dde8baR386).
 	/// The problem is that insertion complexity then depends on the size of the existing queue,
 	/// which makes determining weights hard and could lead to issues like overweight blocks (at
```
Comment (bkchr) [inline polkadot/runtime/parachains/src/assigner_coretime/mod.rs:391]: Implementation details should not be discussed in the documentation of the function. What does the user gets from this? 

If you really want to keep this, it should be a code comment, but not sure there is such a big value in this.

File: polkadot/runtime/parachains/src/assigner_coretime/mod.rs [hunk 2]
```diff
@@ -387,67 +380,56 @@ impl<T: Config> Pallet<T> {
...
+								debug_assert!(schedule.next_schedule.is_none(), "queue.end was supposed to be the end, so the next item must be `None`!");
+								schedule.next_schedule = Some(begin);
+							} else {
+								defensive!("Queue end entry does not exist?");
+							}
+						});
+					}
+
+					CoreSchedules::<T>::mutate((begin, core_idx), |schedule| {
+						let assignments = if let Some(mut old_schedule) = mem::take(schedule) {
```
Comment (bkchr) [inline polkadot/runtime/parachains/src/assigner_coretime/mod.rs:425]:
```suggestion
						let assignments = if let Some(mut old_schedule) = schedule.take() {
```

File: polkadot/runtime/parachains/src/assigner_coretime/mod.rs [hunk 3]
```diff
@@ -387,67 +380,56 @@ impl<T: Config> Pallet<T> {
...
+							old_schedule.assignments.append(&mut assignments);
+							mem::take(&mut old_schedule.assignments)
```
Comment (bkchr) [inline polkadot/runtime/parachains/src/assigner_coretime/mod.rs:427]:
```suggestion
							old_schedule.assignments
```

File: polkadot/runtime/parachains/src/assigner_coretime/mod.rs [hunk 4]
```diff
@@ -387,67 +380,56 @@ impl<T: Config> Pallet<T> {
...
-		})?;
-
-		// Check that the total parts between all assignments are equal to 57600
-		let parts_sum = assignments
-			.iter()
-			.map(|assignment| assignment.1)
-			.try_fold(PartsOf57600::ZERO, |sum, parts| {
-				sum.checked_add(parts).ok_or(Error::<T>::OverScheduled)
-			})?;
-		ensure!(parts_sum.is_full(), Error::<T>::UnderScheduled);
```
Comment (bkchr) [inline polkadot/runtime/parachains/src/assigner_coretime/mod.rs:426]: What happens if the schedule is split up, but we only receive one message in time? Will it break any assumptions if we don't have the full parts scheduled?

**Review by bkchr — APPROVED** [review]

File: polkadot/runtime/parachains/src/assigner_coretime/mod.rs (hunk 4 above)
Comment (eskimor) [reply polkadot/runtime/parachains/src/assigner_coretime/mod.rs:426]: No. I checked, the code does not really care whether the schedule is full or not.

File: polkadot/runtime/parachains/src/assigner_coretime/mod.rs [hunk 5]
```diff
@@ -387,67 +379,56 @@ impl<T: Config> Pallet<T> {
...
-				Ok(Some(cur))
-			}
-		})?;
-
-		// Check that the total parts between all assignments are equal to 57600
-		let parts_sum = assignments
-			.iter()
-			.map(|assignment| assignment.1)
-			.try_fold(PartsOf57600::ZERO, |sum, parts| {
-				sum.checked_add(parts).ok_or(Error::<T>::OverScheduled)
```
Comment (alindima) [inline polkadot/runtime/parachains/src/assigner_coretime/mod.rs:424]: isn't the OverScheduled error still worth checking for?

**Review by alindima — APPROVED** [review]

File: polkadot/runtime/parachains/src/assigner_coretime/mod.rs (hunk 5 above)
Comment (eskimor) [reply polkadot/runtime/parachains/src/assigner_coretime/mod.rs:424]: Something to be checked by the Coretime chain. Erroring out here is not of much help. The scheduling logic does not really care either.

**Pull request merged**